File: backend/app/schemas/engagement.py
"""Engagement request/response schemas (Phase C)."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import FollowTargetType
from app.schemas.public import ArticleCardOut


# ------------------------------------------------------------------ beacon --
class BeaconEventIn(BaseModel):
    short_id: str = Field(min_length=4, max_length=12)
    type: str = Field(description="view | read | scroll | share | not_interested")
    value: int | None = Field(
        default=None, ge=0, le=100_000, description="read: seconds · scroll: percent"
    )


class BeaconIn(BaseModel):
    anon_id: str | None = Field(
        default=None,
        max_length=64,
        description="Stable client-generated id for anonymous readers",
    )
    events: list[BeaconEventIn] = Field(min_length=1, max_length=20)


class BeaconOut(BaseModel):
    accepted: int


# ------------------------------------------------------------- interactions --
class EngagementCountsOut(BaseModel):
    like_count: int
    comment_count: int
    share_count: int


class MyArticleFlagsOut(BaseModel):
    liked: bool
    bookmarked: bool


class ReportIn(BaseModel):
    reason: str = Field(description="spam | abuse | misinformation | copyright | other")
    note: str | None = Field(default=None, max_length=500)


# ---------------------------------------------------------------- comments --
class CommentIn(BaseModel):
    body: str = Field(min_length=1, max_length=2000)
    parent_id: int | None = None


class CommentOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    parent_id: int | None
    body: str
    author_name_te: str
    author_name_en: str
    is_mine: bool = False
    created_at: datetime


class CommentListOut(BaseModel):
    total_visible: int
    comments: list[CommentOut]


# ----------------------------------------------------------------- follows --
class FollowIn(BaseModel):
    target_type: FollowTargetType
    slug: str = Field(min_length=1, max_length=120)


class FollowOut(BaseModel):
    target_type: FollowTargetType
    slug: str
    name_te: str
    name_en: str


class FollowListOut(BaseModel):
    follows: list[FollowOut]


class FollowStateOut(BaseModel):
    following: bool


# ------------------------------------------------------------------ library --
class HistoryItemOut(BaseModel):
    article: ArticleCardOut
    seconds: int
    max_scroll_pct: int
    last_read_at: datetime


class HistoryOut(BaseModel):
    items: list[HistoryItemOut]
    next_offset: int | None = None


class BookmarksOut(BaseModel):
    articles: list[ArticleCardOut]
    next_offset: int | None = None


class FollowingFeedOut(BaseModel):
    articles: list[ArticleCardOut]
    next_offset: int | None = None

File: mobile/src/api/engagement.ts
import { api } from './client';
import type { ArticleCard } from './types';

/** Engagement API (Phase C) — the app twin of the web engagement feature. */

export type FollowTargetType = 'category' | 'tag' | 'district' | 'mandal' | 'author';

export interface EngagementCounts {
  like_count: number;
  comment_count: number;
  share_count: number;
}

export interface MyArticleFlags {
  liked: boolean;
  bookmarked: boolean;
}

export interface CommentOut {
  id: number;
  parent_id: number | null;
  body: string;
  author_name_te: string;
  author_name_en: string;
  is_mine: boolean;
  created_at: string;
}

export interface FollowItem {
  target_type: FollowTargetType;
  slug: string;
  name_te: string;
  name_en: string;
}

export interface HistoryItem {
  article: ArticleCard;
  seconds: number;
  max_scroll_pct: number;
  last_read_at: string;
}

export async function setLike(shortId: string, liked: boolean): Promise<EngagementCounts> {
  const method = liked ? 'post' : 'delete';
  const { data } = await api[method]<EngagementCounts>(`/articles/${shortId}/like`);
  return data;
}

export async function setBookmark(shortId: string, bookmarked: boolean): Promise<MyArticleFlags> {
  const method = bookmarked ? 'post' : 'delete';
  const { data } = await api[method]<MyArticleFlags>(`/articles/${shortId}/bookmark`);
  return data;
}

export async function fetchMyFlags(shortId: string): Promise<MyArticleFlags> {
  const { data } = await api.get<MyArticleFlags>(`/articles/${shortId}/me`);
  return data;
}

export async function fetchComments(
  shortId: string,
): Promise<{ total_visible: number; comments: CommentOut[] }> {
  const { data } = await api.get(`/public/articles/${shortId}/comments`);
  return data;
}

export async function addComment(shortId: string, body: string): Promise<CommentOut> {
  const { data } = await api.post<CommentOut>(`/articles/${shortId}/comments`, {
    body,
    parent_id: null,
  });
  return data;
}

export async function deleteComment(commentId: number): Promise<void> {
  await api.delete(`/comments/${commentId}`);
}

export async function reportArticle(shortId: string, reason: string): Promise<void> {
  await api.post(`/articles/${shortId}/report`, { reason, note: null });
}

export async function setFollow(
  targetType: FollowTargetType,
  slug: string,
  following: boolean,
): Promise<void> {
  if (following) {
    await api.post('/follow', { target_type: targetType, slug });
  } else {
    await api.delete('/follow', { data: { target_type: targetType, slug } });
  }
}

export async function fetchMyFollows(): Promise<FollowItem[]> {
  const { data } = await api.get<{ follows: FollowItem[] }>('/users/me/follows');
  return data.follows;
}

export async function fetchBookmarks(
  offset = 0,
): Promise<{ articles: ArticleCard[]; next_offset: number | null }> {
  const { data } = await api.get('/users/me/bookmarks', { params: { offset, limit: 20 } });
  return data;
}

export async function fetchHistory(
  offset = 0,
): Promise<{ items: HistoryItem[]; next_offset: number | null }> {
  const { data } = await api.get('/users/me/history', { params: { offset, limit: 20 } });
  return data;
}

export async function fetchFollowingFeed(
  offset = 0,
): Promise<{ articles: ArticleCard[]; next_offset: number | null }> {
  const { data } = await api.get('/users/me/following', { params: { offset, limit: 20 } });
  return data;
}

export async function fetchForYou(
  offset = 0,
  limit = 20,
): Promise<{ articles: ArticleCard[]; next_offset: number | null }> {
  const { data } = await api.get('/users/me/for-you', { params: { offset, limit } });
  return data;
}
